packages/tpetra/core/example/Finite-Element-Assembly/fem_assembly_type3_staticprofile.cpp

```diff
@@ -0,0 +1,322 @@
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+// @HEADER
+#include <cmath>
+#include <iostream>
+#include <iomanip>
+#include <sstream>
+
+#include <Tpetra_DefaultPlatform.hpp>
+#include <Tpetra_Version.hpp>
+#include <MatrixMarket_Tpetra.hpp>
+#include <Teuchos_RCP.hpp>
+#include <Teuchos_GlobalMPISession.hpp>
+#include <Teuchos_FancyOStream.hpp>
+
+#include "typedefs.hpp"
+#include "MeshDatabase.hpp"
+#include "Element.hpp"
+
+#define PRINT_VERBOSE 0
+
+
+
+int main (int argc, char *argv[]) 
+{
+  using Teuchos::RCP;
+  using Teuchos::TimeMonitor;
+
+  const GlobalOrdinal GO_INVALID = Teuchos::OrdinalTraits<GlobalOrdinal>::invalid();
+
+  auto out = Teuchos::getFancyOStream (Teuchos::rcpFromRef (std::cout));
+  
+  // MPI boilerplate
+  Teuchos::GlobalMPISession mpiSession (&argc, &argv, NULL);
+  RCP<const Teuchos::Comm<int> > comm = Tpetra::DefaultPlatform::getDefaultPlatform ().getComm();
+
+  // Initialize Kokkos
+  Kokkos::initialize();
+
+  // Processor decomp (only works on perfect squares)
+  int numProcs  = comm->getSize();
+  int sqrtProcs = sqrt(numProcs); 
+
+  if(sqrtProcs*sqrtProcs != numProcs) 
+  {
+    if(0 == mpiSession.getRank())
+      std::cerr << "Error: Invalid number of processors provided, num processors must be a perfect square." << std::endl;
+    return -1;
+  }
+  int procx = sqrtProcs;
+  int procy = sqrtProcs;
+
+  // Generate a simple 3x3 mesh
+  int nex = 3;
+  int ney = 3;
+  MeshDatabase mesh(comm,nex,ney,procx,procy);
+  mesh.print(std::cout);
+
+  int maxEntriesPerRow = 16;
+
+  // Build Tpetra Maps
+  // -----------------
+  // -- https://trilinos.org/docs/dev/packages/tpetra/doc/html/classTpetra_1_1Map.html#a24490b938e94f8d4f31b6c0e4fc0ff77
+  RCP<const MapType> row_map = rcp(new MapType(GO_INVALID, mesh.getOwnedNodeGlobalIDs(), 0, comm));
+
+  #if PRINT_VERBOSE
+  row_map->describe(*out);
+  #endif
+
+  // Type-3: Graph Construction
+  // --------------------------
+  // - Loop over every element in the mesh.
+  //   - Get list of nodes associated with each element.
+  //   - Insert node contributions if the node (row) is owned locally.
+  //   
+  auto domain_map = row_map;
+  auto range_map  = row_map;
+
+  auto owned_element_to_node_ids = mesh.getOwnedElementToNode();
+  auto ghost_element_to_node_ids = mesh.getGhostElementToNode();
+
+  RCP<TimeMonitor> timerGlobal = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("X) Global")));
+  RCP<TimeMonitor> timerElementLoopGraph = rcp(new TimeMonitor(*TimeMonitor::getNewTimer("1) ElementLoop  (Graph)")));
+
+  RCP<GraphType> crs_graph = rcp(new GraphType(row_map, maxEntriesPerRow, Tpetra::StaticProfile));
+  
+  // Using 4 because we're using quads for this example, so there will be 4 nodes associated with each element.
+  Teuchos::Array<GlobalOrdinal> global_ids_in_row(4);
```

Comment:
We want to move away from `Teuchos::Array` in favor of native C++ data structures like `std::vector`.  See note below on `insertGlobalIndices`.